cli/src/generate/templates/Makefile

```diff
@@ -0,0 +1,113 @@
+
+ifeq (, $(PARSER_URL))
+	PARSER_URL := $(subst .git,,$(PARSER_REPO_URL))
+ifeq ($(shell echo $(PARSER_URL) | grep '^[a-z][-+.0-9a-z]*://'),)
+	PARSER_URL := $(subst :,/,$(PARSER_URL))
+	PARSER_URL := $(subst git@,https://,$(PARSER_URL))
+endif
+endif
+
+# install directory layout
+PREFIX ?= /usr/local
+INCLUDEDIR ?= $(PREFIX)/include
+LIBDIR ?= $(PREFIX)/lib
+PCLIBDIR ?= $(LIBDIR)/pkgconfig
+
+# collect C++ sources, and link if necessary
+CPPSRC := $(wildcard $(SRC_DIR)/*.cc)
+
+ifeq (, $(CPPSRC))
+	ADDITIONALLIBS :=
+else
+	ADDITIONALLIBS := -lc++
+endif
+
+# collect sources
+SRC := $(wildcard $(SRC_DIR)/*.c)
+SRC += $(CPPSRC)
+OBJ := $(addsuffix .o,$(basename $(SRC)))
+
+# ABI versioning
+SONAME_MAJOR := 0
+SONAME_MINOR := 0
+
+CFLAGS ?= -O3 -Wall -Wextra -I$(SRC_DIR)
+CXXFLAGS ?= -O3 -Wall -Wextra -I$(SRC_DIR)
+override CFLAGS += -std=gnu99 -fPIC
+override CXXFLAGS += -fPIC
+
+# OS-specific bits
+ifeq ($(shell uname),Darwin)
+	SOEXT = dylib
+	SOEXTVER_MAJOR = $(SONAME_MAJOR).dylib
+	SOEXTVER = $(SONAME_MAJOR).$(SONAME_MINOR).dylib
+	LINKSHARED := $(LINKSHARED)-dynamiclib -Wl,
+	ifneq ($(ADDITIONALLIBS),)
+	LINKSHARED := $(LINKSHARED)$(ADDITIONALLIBS),
+	endif
+	LINKSHARED := $(LINKSHARED)-install_name,$(LIBDIR)/libtree-sitter-$(LANGUAGE_NAME).$(SONAME_MAJOR).dylib,-rpath,@executable_path/../Frameworks
+else ifneq (,$(filter $(shell uname),Linux FreeBSD NetBSD DragonFly))
+	SOEXT = so
+	SOEXTVER_MAJOR = so.$(SONAME_MAJOR)
+	SOEXTVER = so.$(SONAME_MAJOR).$(SONAME_MINOR)
+	LINKSHARED := $(LINKSHARED)-shared -Wl,
+	ifneq ($(ADDITIONALLIBS),)
+	LINKSHARED := $(LINKSHARED)$(ADDITIONALLIBS)
+	endif
+	LINKSHARED := $(LINKSHARED)-soname,libtree-sitter-$(LANGUAGE_NAME).so.$(SONAME_MAJOR)
+else ifeq ($(OS),Windows_NT)
+	$(error "Windows is not supported")
+endif
+ifneq (,$(filter $(shell uname),FreeBSD NetBSD DragonFly))
+	PCLIBDIR := $(PREFIX)/libdata/pkgconfig
+endif
+
+all: libtree-sitter-$(LANGUAGE_NAME).a libtree-sitter-$(LANGUAGE_NAME).$(SOEXTVER) bindings/c/$(LANGUAGE_NAME).h bindings/c/tree-sitter-$(LANGUAGE_NAME).pc
+
+libtree-sitter-$(LANGUAGE_NAME).a: $(OBJ)
+	$(AR) rcs $@ $^
+
+libtree-sitter-$(LANGUAGE_NAME).$(SOEXTVER): $(OBJ)
+	$(CC) $(LDFLAGS) $(LINKSHARED) $^ $(LDLIBS) -o $@
+	ln -sf $@ libtree-sitter-$(LANGUAGE_NAME).$(SOEXT)
+	ln -sf $@ libtree-sitter-$(LANGUAGE_NAME).$(SOEXTVER_MAJOR)
+
+bindings/c/$(LANGUAGE_NAME).h:
+	sed -e 's|@UPPER_LANGUAGENAME@|$(UPPER_LANGUAGE_NAME)|' \
+		-e 's|@LANGUAGENAME@|$(LANGUAGE_NAME)|' \
+		bindings/c/tree-sitter.h.in > $@
+
+bindings/c/tree-sitter-$(LANGUAGE_NAME).pc:
+	sed -e 's|@LIBDIR@|$(LIBDIR)|;s|@INCLUDEDIR@|$(INCLUDEDIR)|;s|@VERSION@|$(VERSION)|' \
+		-e 's|=$(PREFIX)|=$${prefix}|' \
+		-e 's|@PREFIX@|$(PREFIX)|' \
+		-e 's|@ADDITIONALLIBS@|$(ADDITIONALLIBS)|' \
+		-e 's|@LANGUAGENAME@|$(LANGUAGE_NAME)|' \
+		-e 's|@PARSERURL@|$(PARSER_URL)|' \
+		bindings/c/tree-sitter.pc.in > $@
+
+install: all
+	install -d '$(DESTDIR)$(LIBDIR)'
```

Comment:
Is this the best place to install parser libraries, or would they better belong in a `tree-sitter` subdirectory?

Is my understanding correct that the parser libraries aren't stand-alone libraries that can be linked by any program, but instead have to be loaded by calling the appropriate function in `libtree-sitter`?  If so, I think that means they probably don't belong on in the default system library path.

Reply:
They can be used as a standalone libraries that provide a parser data and a lexing logic for the parser core from `libtree-sitter.so` but if to install grammars to a system path as system libraries it's required to follow a lib names convention and name parser libs like `libtree-sitter-c.so`.

Reply:
> Is this the best place to install parser libraries, or would they better belong in a `tree-sitter` subdirectory?

That's not a bad idea (GStreamer plugins for one also live in `/usr/lib/gstreamer-1.0`),
but distros like [Arch] and [Alpine] already have some parser libraries directly in `/usr/lib`.

[Alpine]: https://pkgs.alpinelinux.org/contents?branch=edge&name=tree-sitter-bash&arch=x86_64&repo=community
[Arch]: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=tree-sitter-bash